### PR TITLE
[Helm chart] Refactoring scheduling methods

### PR DESCRIPTION
**Pull request by nlamirault**
**Description**: Currently we can't customize affinity, nodeSelectors or tolerations.
This makes it impossible to deploy OSM on an ARM64 cluster.

https://github.com/openservicemesh/osm/issues/4702

**Testing done**:

**Affected area**:
| Functional Area            |     |
| -------------------------- | --- |
| New Functionality          | [ ] |
| CI System                  | [ ] |
| CLI Tool                   | [ ] |
| Certificate Management     | [ ] |
| Control Plane              | [ ] |
| Demo                       | [ ] |
| Documentation              | [ ] |
| Egress                     | [ ] |
| Ingress                    | [ ] |
| Install                    | [ ] |
| Networking                 | [ ] |
| Observability              | [ ] |
| Performance                | [ ] |
| SMI Policy                 | [ ] |
| Security                   | [ ] |
| Sidecar Injection          | [ ] |
| Tests                      | [ ] |
| Upgrade                    | [ ] |
| Other                      | [x ] |

Please answer the following questions with yes/no.

1. Does this change contain code from or inspired by another project?
    -   Did you notify the maintainers and provide attribution?

2. Is this a breaking change?

3. Has documentation corresponding to this change been updated in the [osm-docs](https://github.com/openservicemesh/osm-docs) repo (if applicable)?
Yes into the chart README.

### CODE REVIEW

File: charts/osm/values.yaml [hunk 1]
```diff
@@ -166,13 +229,16 @@ osm:
   enableReconciler: false
 
   # -- Deploy Prometheus with OSM installation
-  deployPrometheus: false
+  deployPrometheus: true
```
Comment (nshankar13) [inline charts/osm/values.yaml:232]: What's the reasoning for changing the default values for deploying Prometheus, Grafana, Jaeger, and fluentbit?

**Comment by nshankar13** [comment]
@nlamirault can you also share the full `osm install` command and CLI output on ARM64 cluster/environment, and if possible also on AMD64?

File: charts/osm/values.yaml (hunk 1 above)
Comment (nlamirault) [reply charts/osm/values.yaml]: A mistake. Fixed

**Comment by nlamirault** [comment]
@nshankar13 can i only generate manifests from CLI ?

**Review by shashankram — COMMENTED** [review]
Thanks @nlamirault for this change. Could you address the following comments:
1. Describe in the PR description what additional testing was done 
2. Update the [release notes](https://github.com/openservicemesh/osm/blob/main/docs/release_notes.md) to reflect the breaking changes to existing Helm chart values.

**Comment by nshankar13** [comment]
@nlamirault manifests/configs not needed, just need to ensure that installation succeeded and that all pods in OSM release namespace are ready.

**Comment by nlamirault** [comment]
My tests was : 

```
NAME                              READY   STATUS    RESTARTS   AGE
osm-injector-7ccd8c9d98-xdmqg     0/1     Pending   0          8d
osm-controller-687f75c598-l4lsd   0/1     Pending   0          8d
osm-bootstrap-5df4b86cc7-wpsbp    0/1     Pending   0          8d
```

kubectl edit to change amd64 to arm64 : 

```
❯ kubectl -n osm-system get pod
NAME                              READY   STATUS    RESTARTS   AGE
osm-bootstrap-74fdcc5589-58f9q    1/1     Running   0          76s
osm-controller-7fc5c87bf4-jd5zp   1/1     Running   0          43s
osm-injector-6dc5fd7658-bqm6q     1/1     Running   0          61s
```

Then update the chart, and check differences :

with https://github.com/Wilfred/difftastic :

```
❯ difft amd64.yaml arm64.yaml
arm64.yaml --- YAML
No syntactic changes.
```

with diff :

```
✖ diff amd64.yaml arm64.yaml
275c275
<                 - amd64
---
>                 - arm64
396c396
<                 - amd64
---
>                 - arm64
515c515
<                 - amd64
---
>                 - arm64
757c757
<         kubernetes.io/arch: amd64
---
>         kubernetes.io/arch: arm64

```

**Comment by nlamirault** [comment]
```
❯ helm install test-osm -f ~/values-arm64.yaml -n test-osm Forks/osm/charts/osm/
NAME: test-osm
LAST DEPLOYED: Thu May  5 09:01:04 2022
NAMESPACE: test-osm
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
Congratulations! The OSM control plane has been installed in your Kubernetes cluster!

❯ kubectl -n test-osm get pods
NAME                              READY   STATUS    RESTARTS   AGE
osm-bootstrap-7db9ccf4b6-d7sl4    1/1     Running   0          2m
osm-controller-7fbf694685-t6t4n   1/1     Running   0          2m1s
osm-injector-c9795d495-wxlc9      1/1     Running   0          2m
```

**Comment by nlamirault** [comment]
@shashankram how i update the release note ? 
Last release is on `v1.1.0` (https://github.com/openservicemesh/osm/blob/main/docs/release_notes.md#release-v110=), but there is a [v1.1.1](https://github.com/openservicemesh/osm/releases/tag/v1.1.1) 
Do i have to create a v1.2.0 on the release note document ?

**Comment by shashankram** [comment]
> @shashankram how i update the release note ? Last release is on `v1.1.0` (https://github.com/openservicemesh/osm/blob/main/docs/release_notes.md#release-v110=), but there is a [v1.1.1](https://github.com/openservicemesh/osm/releases/tag/v1.1.1) Do i have to create a v1.2.0 on the release note document ?

@nlamirault You need to make changes to that file by adding a section for v1.2.0 and listing the breaking changes.

File: charts/osm/values.schema.json [hunk 2]
```diff
@@ -1384,6 +1447,30 @@
...
                     },
                     "additionalProperties": false
+                },
+                "preinstall": {
+                    "$id": "#/properties/osm/properties/preinstall",
+                    "type": "object",
+                    "title": "The preinstall schema",
+                    "description": "Preinstall configurations",
+                    "required": [],
+                    "properties": {
```
Comment (nojnhuh) [inline charts/osm/values.schema.json:1457]: Let's also add `"additionalProperties": false` to these to match everything else.

**Review by nojnhuh — CHANGES_REQUESTED** [review]
Thanks for this change @nlamirault! Left a few comments inline.

File: charts/osm/values.yaml [hunk 3]
```diff
@@ -88,6 +88,41 @@ osm:
...
+          nodeSelectorTerms:
+          - matchExpressions:
+            - key: kubernetes.io/os
+              operator: In
+              values:
+              - linux
+            - key: kubernetes.io/arch
+              operator: In
+              values:
+              - amd64
```
[the remark this answers is not in the document]
Comment (nlamirault) [reply charts/osm/values.yaml:109]: Done. I add  `arm64` in others podAntiAffinity alsoe.

File: charts/osm/values.yaml [hunk 4]
```diff
@@ -108,6 +143,20 @@ osm:
     # -- Image used for Prometheus
     image: prom/prometheus:v2.34.0
 
+    ## Node labels for pod assignment
+    ## Ref: https://kubernetes.io/docs/user-guide/node-selection/
+    nodeSelector: #{}
```
[the remark this answers is not in the document]
Comment (nlamirault) [reply charts/osm/values.yaml:149]: Done.

File: charts/osm/values.yaml [hunk 5]
```diff
@@ -108,6 +143,20 @@ osm:
...
+      kubernetes.io/arch: amd64
+      kubernetes.io/os: linux
+
+    ## Affinity settings for pod assignment
+    ## Ref: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/
+    affinity: {}
```
[the remark this answers is not in the document]
Comment (nlamirault) [reply charts/osm/values.yaml:155]: So only : 
```yaml
nodeSelector:
      kubernetes.io/arch: amd64
```
I think Prometheus images are multi arch now.

[the remark this answers is not in the document]
Comment (nojnhuh) [reply charts/osm/values.yaml:155]: I mean let's define all of the architectures supported by the Prometheus image in the `nodeAffinity` here and remove the `nodeSelector`.

File: charts/osm/values.yaml [hunk 6]
```diff
@@ -246,6 +310,20 @@ osm:
...
 
+    ## Node labels for pod assignment
+    ## Ref: https://kubernetes.io/docs/user-guide/node-selection/
+    nodeSelector:
+      kubernetes.io/arch: amd64
+      kubernetes.io/os: linux
+
+    ## Affinity settings for pod assignment
+    ## Ref: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/
+    affinity: {}
```
Comment (nojnhuh) [inline charts/osm/values.yaml:321]: Let's do the same as Prometheus here and define all of the architectures supported by the image here in `nodeAffinity`.

File: charts/osm/values.yaml [hunk 7]
```diff
@@ -330,9 +444,19 @@ osm:
     # -- Log level for the multicluster gateway
     gatewayLogLevel: error
 
-  # -- Node tolerations applied to control plane pods.
-  # The specified tolerations allow pods to schedule onto nodes with matching taints.
-  controlPlaneTolerations: []
+    ## Node labels for pod assignment
+    ## Ref: https://kubernetes.io/docs/user-guide/node-selection/
+    nodeSelector:
+      kubernetes.io/arch: amd64
```
Comment (nojnhuh) [inline charts/osm/values.yaml:450]: @shashankram The envoy-alpine image is only built for amd64. Do you see any issues if we specify `envoyproxy/envoy` as the default sidecar image instead which also supports arm64?

File: charts/osm/values.yaml [hunk 8]
```diff
@@ -350,12 +474,93 @@ osm:
...
+
+    # -- Node tolerations applied to control plane pods.
+    # The specified tolerations allow pods to schedule onto nodes with matching taints.
+    tolerations: []
+
+  cleanup:
+    ## Node labels for pod assignment
+    ## Ref: https://kubernetes.io/docs/user-guide/node-selection/
+    nodeSelector:
+      kubernetes.io/arch: amd64
```
Comment (nojnhuh) [inline charts/osm/values.yaml:551]: This image also supports arm64, so let's define both amd64 and arm64 as acceptable in the `nodeAffinity`.

**Comment by shalier** [comment]
PR also addresses https://github.com/openservicemesh/osm/issues/4496

**Comment by shalier** [comment]
@nlamirault Please let use know if we can provide more information on the outstanding comments!

**Comment by shalier** [comment]
Created https://github.com/openservicemesh/osm/pull/4842 to address comments in this PR - closing this PR

**Pull request closed without merging**